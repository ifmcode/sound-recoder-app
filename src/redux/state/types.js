// Types to identify dispatch action
const RecordingReducerType = {
  GET_RECORDINGS_SUCCESS: 'GET_RECORDINGS_SUCCESS',
  CREATE_RECORDING_SUCCESS: 'CREATE_RECORDING_SUCCESS',
  UPDATE_RECORDING_SUCCESS: 'UPDATE_RECORDING_SUCCESS',
  REMOVE_RECORDING_SUCCESS: 'REMOVE_RECORDING_SUCCESS',
  RECORDING_OPERATION_FAILED: 'RECORDING_OPERATION_FAILED',
}

// Store state status
const stateStatus = {
  ZERO: 'zero',
  SUCCESS: 'success',
  FAILED: 'failed',
};

export {
  RecordingReducerType,
  stateStatus,
}
