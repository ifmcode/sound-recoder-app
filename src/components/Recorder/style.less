@import (reference) "../../styles/colors.less";
@import (reference) "../../styles/sizes.less";

main {
  .buttons-section, #recording-list-container {
    margin-bottom: 30px;
  }

  .buttons-section, .big-buttons-section {
    position: relative;
    display: flex;
    justify-content: space-between;
  }


  #recording-list-container {
    position: relative;
  }

  .big-buttons-section {
    align-items: center;
    margin-bottom: 50px;
    .button-container {
      span {
        position: relative;
        display: block;
        font-size: @font-size_10;
        line-height: @font-size_10;
        text-transform: uppercase;
        color: @black;
        font-weight: bold;
        margin-bottom: 15px;
        text-align: center;
      }

      .circular-button {
        position: relative;
        display: block;
      }
    }

    span.is-recording {
      color: @bluegreen;
    }
  }

  .playbar-section {
    .rhap_container {
      background-color: transparent;
      padding: 0;
    }

    .rhap_current-time, .rhap_total-time {
      color: @bluegreen;
      height: 20px;
    }

    .rhap_progress-bar {
      background-color: @black;
      height: 16px;
      border-radius: 10px;
      border: 3px solid @darkgrey;
      box-sizing: border-box;
      .rhap_progress-filled {
        border-radius: 10px;
      }
    }

    .rhap_progress-filled {
      background-color: @bluegreen;
    }

    .rhap_controls-section {
      justify-content: center;
      margin: 20px auto;
      width: 67px;
    }

    .rhap_progress-indicator, .rhap_main-controls, .rhap_additional-controls, .rhap_volume-controls {
      display: none;
    }
  }
}