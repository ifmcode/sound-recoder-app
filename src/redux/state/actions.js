import {RecordingReducerType} from './types';
import {getSavedRecordings, saveNewRecording, updateRecording, removeRecording}
  from '../../firebase/firebaseOperations';

/*
  Actions are used by the front to save or restore data from store
*/

const getRecordingsAction = () => async (dispatch) => {
  try {
    const recordings = await getSavedRecordings();
    dispatch({
      type: RecordingReducerType.GET_RECORDINGS_SUCCESS,
      payload: {
        recordings,
      }
    });
  } catch (error) {
    dispatchError(dispatch, error);
  }
}

const createRecordingAction = (newRecording) => async (dispatch) => {
  try {
    await saveNewRecording(newRecording);
    dispatch({
      type: RecordingReducerType.CREATE_RECORDING_SUCCESS,
      payload: {
        newRecording,
      }
    });
  } catch (error) {
    dispatchError(dispatch, error);
  }
}

const updateRecordingAction = (recordingUpdated) => async (dispatch) => {
  try {
    await updateRecording(recordingUpdated);
    dispatch({
      type: RecordingReducerType.UPDATE_RECORDING_SUCCESS,
      payload: {
        recordingUpdated,
      }
    });
  } catch (error) {
    dispatchError(dispatch, error);
  }
}

const removeRecordingAction = (recordingId) => async (dispatch) => {
  try {
    await removeRecording(recordingId);
    dispatch({
      type: RecordingReducerType.REMOVE_RECORDING_SUCCESS,
      payload: {
        recordingId,
      }
    });
  } catch (error) {
    dispatchError(dispatch, error);
  }
}

const dispatchError = (dispatch, error) => {
  dispatch({
    type: RecordingReducerType.RECORDING_OPERATION_FAILED,
    payload: {
      error: error,
    }
  });
}

export {
  getRecordingsAction,
  createRecordingAction,
  updateRecordingAction,
  removeRecordingAction,
}