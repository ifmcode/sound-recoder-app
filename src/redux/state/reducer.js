import {RecordingReducerType, stateStatus} from './types';

const initialData = {
  recordings: [],
  status: stateStatus.ZERO,
  error: null,
}

// Map that identify which function it calls when reducer receive a dispatch action
const ACTION_TYPE_FUNCTION_MAP = {
  [RecordingReducerType.GET_RECORDINGS_SUCCESS]: onGetRecordingsSuccess,
  [RecordingReducerType.CREATE_RECORDING_SUCCESS]: onCreateRecordingSuccess,
  [RecordingReducerType.UPDATE_RECORDING_SUCCESS]: onUpdateRecordingSuccess,
  [RecordingReducerType.REMOVE_RECORDING_SUCCESS]: onRemoveRecordingSuccess,
  [RecordingReducerType.RECORDING_OPERATION_FAILED]: onRecordingOperationFailed,
}

function onGetRecordingsSuccess(state, action) {
  return {
    ...state,
    recordings: action.payload.recordings,
    status: stateStatus.SUCCESS,
  };
}

function onCreateRecordingSuccess(state, action) {
  return {
    ...state,
    recordings: [...state.recordings, action.payload.newRecording],
    status: stateStatus.SUCCESS,
  };
}

function onUpdateRecordingSuccess(state, action) {
  const recordings = state.recordings,
    recordingUpdated = action.payload.recordingUpdated,
    indexOfRecordingtoUpdate = recordings.findIndex((recording) => recording.id === recordingUpdated.id);
  recordings[indexOfRecordingtoUpdate] = recordingUpdated;
  return {
    ...state,
    recordings,
    status: stateStatus.SUCCESS,
  };
}

function onRemoveRecordingSuccess(state, action) {
  const currentRecordings = state.recordings,
    recordingIdToDelete = action.payload.recordingId,
    recordings = currentRecordings.filter((recording) => recording.id !== recordingIdToDelete);
  return {
    ...state,
    recordings,
    status: stateStatus.SUCCESS,
  };
}

function onRecordingOperationFailed(state, action) {
  return {
    ...state,
    error: action.payload.error,
    status: stateStatus.FAILED,
  }
}

function soundRecordingReducer(state = initialData, action) {
  const fn = ACTION_TYPE_FUNCTION_MAP[action.type];
  return (fn)
    ? fn(state, action)
    : state;
}

export {
  soundRecordingReducer,
}
