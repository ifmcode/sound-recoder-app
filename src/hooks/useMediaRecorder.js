import {useMemo, useState} from "react";
import { useDispatch } from "react-redux";
import {createRecordingAction} from '../redux/state/actions';

/**
 * This media recorder hook keeps all the logic needed to perform an audio recording decoupling it from any component
 * so it can be reused if needed.
 *
 * This way, when it comes to testing you only need to mock this `recorder` hook and its returned values.
 *
 * The hook in use would look like this:
 * const { recording, isRecording } = useMediaRecorded(stream);
 *
 *
 * @param {Object} stream
 * @returns {{
 *  recorder: { start: Function, stop: Function },
 *  isRecording: boolean,
 * }}
 *
 */
export default function useMediaRecorder(stream) {
  const [isRecording, setIsRecording] = useState(false),
    [chunks, setChunks] = useState([]),
    recorder = useMemo(() => new MediaRecorder(stream), [stream]),
    dispatch = useDispatch();

  const handleStart = () => {
    setIsRecording(true);
  }

  const handleStop = async () => {
    const blob = new Blob(chunks, { 'type' : 'audio/ogg; codecs=opus' }),
      audioURL = window.URL.createObjectURL(blob),
      now = new Date();

    const newRecording = {
      stream: audioURL,
      name: now.toISOString().split('.')[0].split('T').join(' '),
      id: `id${window.performance.now().toString()}`
    };
    dispatch(createRecordingAction(newRecording));
    setChunks([]);
    setIsRecording(false);
  }

  const handleDataAvailable = (e) => {
    setChunks(currentChunks => [...currentChunks, e.data]);
  }

  recorder.onstop = handleStop;
  recorder.onstart = handleStart;
  recorder.ondataavailable = handleDataAvailable;

  return {
    recorder,
    isRecording,
  }
}
