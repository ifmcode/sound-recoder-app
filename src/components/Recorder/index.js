import './style.less'
import 'react-h5-audio-player/lib/styles.less'
import { useEffect, useState, useRef } from 'react'
import {useDispatch, useSelector} from 'react-redux'
import PropTypes from 'prop-types'
import useMediaRecorder from "../../hooks/useMediaRecorder";
import {getRecordingsAction, updateRecordingAction, removeRecordingAction} from '../../redux/state/actions';
import {stateStatus} from '../../redux/state/types';
import RecordingList from '../RecordingList';
import Button from '../Button';
import CircularButton from '../CircularButton';
import AudioPlayer from 'react-h5-audio-player';
import Popup from '../Popup';
import { PopupType } from '../Popup/popupConstants'

/**
 * Reducer component
 * @param {MediaStream} stream
 * @returns 
 */
const Recorder = ({ stream }) => {
  const dispatch = useDispatch(),
    recordingsStore = useSelector(store => store.soundRecordingReducer),
    { recorder, isRecording } = useMediaRecorder(stream),
    [ selectedRecording, setSelectedRecording] = useState(null),
    [isPlaying, setIsPlaying] = useState(false),
    [advice, setAdvice] = useState(""),
    audioPlayer = useRef(),
    [popupConfiguration, setPopupConfiguration] = useState({
      visible: false,
      popupType: null,
      onConfirm: null,
    });
  
  // Hook to control the store status, depending of that, request the recordings or set the selected recording
  useEffect(() => {
    if (recordingsStore.status === stateStatus.ZERO) {
      dispatch(getRecordingsAction());
    } else if (recordingsStore.status === stateStatus.SUCCESS) {
      const recordings = recordingsStore.recordings;
      if (recordings?.length) {
        setSelectedRecording(recordings[recordings.length - 1]);
      }
    }
  }, [recordingsStore.status, recordingsStore.recordings, dispatch]);

  const toggleRecording = () => {
    if (!isRecording) {
      recorder.start(1000);
    } else {
      recorder.stop();
    }
  }

  const togglePlayPause = () => {
    if (selectedRecording) {
      const prevStatus = isPlaying;
      setIsPlaying(!prevStatus);
      if (!prevStatus) {
        audioPlayer?.current?.audio?.current?.play();
      } else {
        audioPlayer?.current?.audio?.current?.pause();
      }
    }
  }

  const editRecordingName = (newName) => {
    if (selectedRecording) {
      if (newName === "") {
        setAdvice("Please insert a name");
      } else {
        setAdvice("");
        selectedRecording.name = newName;
        dispatch(updateRecordingAction(selectedRecording));
        hidePopup();
      }
    }
  }

  const deleteRecording = () => {
    if (selectedRecording) {
      dispatch(removeRecordingAction(selectedRecording.id));
      const recordings = recordingsStore.recordings;
      if (recordings.length) {
        setSelectedRecording(recordings[0]);
      } else {
        // The selectedRecording is set to null to prevent it from being used after deletion.
        setSelectedRecording(null);
      }
      hidePopup();
    }
  }

  const hidePopup = () => {
    setPopupConfiguration({
      visible: false
    });
  }

  const onClickDeleteButton = () => {
    if (selectedRecording) {
      // It shows the edit popup
      setPopupConfiguration({
        visible: true,
        popupType: PopupType.DELETE_POPUP,
        onConfirm: deleteRecording,
      });
    }
  }

  const onClickEditButton = () => {
    if (selectedRecording) {
      // It shows the edit popup
      setPopupConfiguration({
        visible: true,
        popupType: PopupType.EDIT_POPUP,
        onConfirm: editRecordingName,
      });
    }
  }

  const onSelectRecording = (recording) => {
    // If the recording selected is clicked second time, will be reproduced.
    if (recording.id !== selectedRecording.id) {
      setSelectedRecording(recording);
    } else {
      togglePlayPause();
    }
  }

  return (
    <>
      {
        popupConfiguration.visible &&
          <Popup
            popupType= {popupConfiguration.popupType}
            selectedRecording= {selectedRecording}
            onConfirm= {popupConfiguration.onConfirm}
            adviceMessage= {advice}
            onDeny= {hidePopup}
          />
      }
      <section className='buttons-section'>
        <Button buttonText="My list"/>
        <CircularButton iconName="settings" iconSize="small"/>
      </section>
      <RecordingList
        recordings= {recordingsStore.recordings}
        selectedRecording= {selectedRecording}
        storeStatus= {recordingsStore.status}
        onSelectRecordingHandler= {onSelectRecording}
      />
      <section className='big-buttons-section'>
        <div className='button-container'>
          <span>Send</span>
          <CircularButton iconName="plane" iconSize="medium"/>
        </div>
        <div className='button-container'>
          <span className={isRecording ? "is-recording" : ""}>Rec</span>
          <CircularButton
            iconName="recording"
            iconSize="big"
            isRecording={isRecording}
            onClick={toggleRecording}
          />
        </div>
        <div className='button-container'>
          <span>Delete</span>
          <CircularButton
            iconName="trash"
            iconSize="medium"
            onClick={onClickDeleteButton}
          />
        </div>
      </section>
      {
        selectedRecording?.stream &&
          <section className='playbar-section'>
            <AudioPlayer
              ref={audioPlayer}
              src={selectedRecording?.stream}
              preload="metadata"
              showJumpControls={false}
              showDownloadProgress={false}
              autoPlayAfterSrcChange={false}
              onPlay={() => setIsPlaying(true)}
              onPause={() => setIsPlaying(false)}
            />
          </section>
      }
      <section className='buttons-section last-section'>
        <Button buttonText="Save" buttonSize="large"/>
        <Button
          buttonText={isPlaying ? "Pause" : "Play"}
          buttonSize="large"
          onClick={() => togglePlayPause()}
        />
        <Button
          buttonText="Edit"
          buttonSize="large"
          onClick={onClickEditButton}
        />
      </section>
    </>
  )
}

Recorder.propTypes = {
  stream: PropTypes.object
};

export default Recorder
