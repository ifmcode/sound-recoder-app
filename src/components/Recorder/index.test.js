import React, {useState} from 'react';
import {fireEvent, render, screen} from '@testing-library/react'
import Recorder from './index'
import useMediaRecorder from "../../hooks/useMediaRecorder";
import { applyMiddleware, combineReducers, createStore } from 'redux';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { soundRecordingReducer } from '../../redux/state/reducer';
import {stateStatus} from '../../redux/state/types';

jest.mock('../../hooks/useMediaRecorder')

/**
 * Following the Object mother pattern we have this small fn that generates a valid object
 * that matches the structure of a recording
 *
 * @param {{ idNumber?: number, name?: string }} options
 * @returns {{ stream: string, name: string, id: string }}
 */
 function createMockRecording({
  idNumber = Math.floor(Math.random() * 100),
  name = new Date().toISOString().split('.')[0].split('T').join(' ')
}) {
  return {
    stream: "audioUrl",
    name,
    id: `id${idNumber}`
  }
}

/**
 * Applying the same pattern above we wrap the previous recording generator
 * to create a random list of recordings
 *
 * @param {number} length
 * @returns {{ stream: string, name: string, id: string }[]}
 */
function createMockRecordingList(length = 10) {
  const emptyList = new Array(length).fill(null);
  return emptyList.map(() => getMockRecording({}))
}

/**
 * We create a mocked version of our hook that will interact with the component in the same exact way
 * and will expose the same API too.
 *
 * This mock is typically placed in the same directory of the original hook within a folder called `__mocks__`
 * keeping the same file name as the original and jest will override the hook functionality automatically.
 * But in this case we would loose the option to pass it a default list of recordings or would be more difficult to do so.
 *
 * @param {{ stream: string, name: string, id: string }[]} defaultRecordings
 * An optional list of default recordings to that we don't need to interact with the component to create a previous list of recordings
 * having also control over the data of each one to be able to assert on the list data.
 *
 * @returns {{
 *  recorder: { stop: Function, start: Function },
 *  isRecording: boolean,
 * }}
 */
const mockUseMediaRecorder = (defaultRecordings = []) => {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const [isRecording, setIsRecording] = useState(false)

  const recorder = {
    start: () => setIsRecording(true),
    stop: () => setIsRecording(false),
  };

  return {recorder, isRecording}
};

describe('Test Recorder with an empty list of recordings', () => {

  let wrapper,
    mockStore;

  beforeEach(() => {
    const reducer = combineReducers({
      soundRecordingReducer,
    });
    mockStore = createStore(reducer, applyMiddleware(thunk));
    useMediaRecorder.mockImplementation(() => mockUseMediaRecorder());

    wrapper = render(
      <Provider store={mockStore}>
        <Recorder stream={{}}/>
      </Provider>
    );
  });
  
  test("should render the Recorder component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('renders without crashing', () => {
    const recordButton = wrapper.container.querySelector('.big-button'),
      playButton = screen.getByRole("button", { name: 'Play' });
    expect(recordButton).toBeInTheDocument();
    expect(playButton).toBeInTheDocument();
    expect(recordButton).toHaveClass('recording-button');
  });

  test('user can start a recording pressing the button', () => {
    const recordButton = wrapper.container.querySelector('.big-button');
    expect(recordButton).toHaveClass('recording-button')
    fireEvent.click(recordButton)
    expect(recordButton).toHaveClass('is-recording')
  });

  test('user can see the loading state of the recorder', () => {
    expect(screen.getByText("- Loading recordings -")).toBeInTheDocument();
  });

  test('user can see the loaded state of the recorder', async () => {
    const state = mockStore.getState();
    state.soundRecordingReducer.status = stateStatus.SUCCESS;
    //expect(await screen.findByText("- There are not recordings -")).toBeInTheDocument();
  });

  test('adds a new recording to the list when the user clicks stop', () => {
    const recordButton = wrapper.container.querySelector('.recording-button');
    fireEvent.click(recordButton)
    expect(recordButton).toHaveClass('is-recording')
    fireEvent.click(recordButton)
    expect(wrapper.container.querySelector(".is-recording")).toBeNull()
    /*const state = mockStore.getState();
    state.soundRecordingReducer.recordings = [getMockRecording({})];
    console.log("|||| AQUI: ", mockStore.getState());
    screen.debug();
    expect(screen.queryAllByRole("listitem")).toHaveLength(1)*/
  })
})
