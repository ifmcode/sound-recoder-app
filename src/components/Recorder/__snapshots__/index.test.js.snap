// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Recorder with an empty list of recordings should render the Recorder component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="buttons-section"
      >
        <button
          class="button undefined-button"
        >
          My list
        </button>
        <button
          class="circular-button settings-button small-button "
        >
          <div
            class="circle-1"
          >
            <div
              class="circle-2"
            >
              <div
                class="icon settings-icon"
              >
                <svg>
                  settings.svg
                </svg>
              </div>
            </div>
          </div>
        </button>
      </section>
      <div
        id="recording-list-container"
      >
        <p
          class="default-phrase-recording-list"
        >
          - Loading recordings -
        </p>
      </div>
      <section
        class="big-buttons-section"
      >
        <div
          class="button-container"
        >
          <span>
            Send
          </span>
          <button
            class="circular-button plane-button medium-button "
          >
            <div
              class="circle-1"
            >
              <div
                class="circle-2"
              >
                <div
                  class="icon plane-icon"
                >
                  <svg>
                    plane.svg
                  </svg>
                </div>
              </div>
            </div>
          </button>
        </div>
        <div
          class="button-container"
        >
          <span
            class=""
          >
            Rec
          </span>
          <button
            class="circular-button recording-button big-button "
          >
            <div
              class="circle-1"
            >
              <div
                class="circle-2"
              >
                <div
                  class="icon recording-icon"
                >
                  <svg>
                    recording.svg
                  </svg>
                </div>
              </div>
            </div>
          </button>
        </div>
        <div
          class="button-container"
        >
          <span>
            Delete
          </span>
          <button
            class="circular-button trash-button medium-button "
          >
            <div
              class="circle-1"
            >
              <div
                class="circle-2"
              >
                <div
                  class="icon trash-icon"
                >
                  <svg>
                    trash.svg
                  </svg>
                </div>
              </div>
            </div>
          </button>
        </div>
      </section>
      <section
        class="buttons-section last-section"
      >
        <button
          class="button large-button"
        >
          Save
        </button>
        <button
          class="button large-button"
        >
          Play
        </button>
        <button
          class="button large-button"
        >
          Edit
        </button>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="buttons-section"
    >
      <button
        class="button undefined-button"
      >
        My list
      </button>
      <button
        class="circular-button settings-button small-button "
      >
        <div
          class="circle-1"
        >
          <div
            class="circle-2"
          >
            <div
              class="icon settings-icon"
            >
              <svg>
                settings.svg
              </svg>
            </div>
          </div>
        </div>
      </button>
    </section>
    <div
      id="recording-list-container"
    >
      <p
        class="default-phrase-recording-list"
      >
        - Loading recordings -
      </p>
    </div>
    <section
      class="big-buttons-section"
    >
      <div
        class="button-container"
      >
        <span>
          Send
        </span>
        <button
          class="circular-button plane-button medium-button "
        >
          <div
            class="circle-1"
          >
            <div
              class="circle-2"
            >
              <div
                class="icon plane-icon"
              >
                <svg>
                  plane.svg
                </svg>
              </div>
            </div>
          </div>
        </button>
      </div>
      <div
        class="button-container"
      >
        <span
          class=""
        >
          Rec
        </span>
        <button
          class="circular-button recording-button big-button "
        >
          <div
            class="circle-1"
          >
            <div
              class="circle-2"
            >
              <div
                class="icon recording-icon"
              >
                <svg>
                  recording.svg
                </svg>
              </div>
            </div>
          </div>
        </button>
      </div>
      <div
        class="button-container"
      >
        <span>
          Delete
        </span>
        <button
          class="circular-button trash-button medium-button "
        >
          <div
            class="circle-1"
          >
            <div
              class="circle-2"
            >
              <div
                class="icon trash-icon"
              >
                <svg>
                  trash.svg
                </svg>
              </div>
            </div>
          </div>
        </button>
      </div>
    </section>
    <section
      class="buttons-section last-section"
    >
      <button
        class="button large-button"
      >
        Save
      </button>
      <button
        class="button large-button"
      >
        Play
      </button>
      <button
        class="button large-button"
      >
        Edit
      </button>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
